mass_mailing_partner/models/mail_mass_mailing_contact.py

```diff
@@ -0,0 +1,87 @@
+# -*- coding: utf-8 -*-
+# See README.rst file on addon root folder for license details
+
+from openerp import models, fields, api, _
+from openerp.exceptions import ValidationError
+
+
+class MailMassMailingContact(models.Model):
+    _inherit = 'mail.mass_mailing.contact'
+
+    partner_id = fields.Many2one(comodel_name='res.partner', string="Partner",
+                                 domain=[('email', '!=', False)])
+
+    @api.one
+    @api.onchange('partner_id')
+    def _onchange_partner(self):
+        if self.partner_id:
+            self.name = self.partner_id.name
+            self.email = self.partner_id.email
+
+    @api.one
+    @api.constrains('partner_id')
+    def _check_partner_id(self):
+        """Ensure at least one name is set."""
+        if self.partner_id:
+            already = self.search([
+                ('id', '!=', self.id),
+                ('partner_id', '=', self.partner_id.id),
+                ('list_id', '=', self.list_id.id),
+            ])
+            if already:
+                raise ValidationError(
+                    _('Partner already exists in this mailing list.'))
+
+    @api.model
+    def create(self, vals):
+        if not vals.get('partner_id'):
+            vals = self._check_partner(vals)
+        vals = self._check_name_email(vals)
+        return super(MailMassMailingContact, self).create(vals)
+
+    @api.one
```

Comment:
This makes mass-write over this model very slow, as the operation is done record by record, and searching again and again for the same partner. You should rewrite it to support @api.multi and make the operations only one time. The only problem here is the mailing_list.partner_mandatory check, that in a multi-record environment, there can be some that has this flag marked and some not. You will have to check if `any(self.mapped('mailing_list.partner_mandatory')` and create the partner in that case.